File: src/utils/insert_cassette_exons.m
% written by Andre Kahles, Mpi Tuebingen, Germany, 2012

function [genes, inserted] = insert_cassette_exons(genes, CFG)
% [genes, inserted] = insert_cassette_exons(genes, CFG)

inserted.cassette_exon = 0 ;

%%% form chunks for quick sorting
chunks = [[genes.chr_num]', cast([genes.strand]', 'int32'), [genes.start]', [genes.stop]'];
[chunks, chunk_idx] = sortrows(chunks) ;
assert(issorted(chunks, 'rows'));

strands = '+-';

%%% form all possible combinations of contigs and strands --> regions
regions = init_regions(CFG.bam_fnames);
%%% keep only chromosomes found in genes
keepidx = find(ismember([regions.chr_num], unique([genes.chr_num])));
regions = regions(keepidx);

c = 1; cov = 0 ;
num_exons_added = 0 ;
num_exons = 0 ; 

for j = 1:length(regions),
	chr = regions(j).chr_num;
	s = find(regions(j).strand == strands);
	
	% fill the chunks on the corresponding chromosome
	while c <= size(chunks,1),
		if chunks(c,1) > chr || chunks(c,2)>strands(s),
			break;
		end ;
		if ~(chunks(c,1) == chr),
			error('c logic seems wrong\n') ;
		end ;

		if CFG.verbose && mod(c, 100) == 0,
			fprintf('\r %i(%i) genes done (found %i new cassette exons in %i tested intron pairs, %2.1f%%)', c, ...
			size(chunks,1), num_exons_added, num_exons, 100*num_exons_added/num_exons);
		end	;

		gg = genes(chunk_idx(c));
		gg.strand = strands(s);
		rm_strands = ~isfield(gg, 'strands');
		gg.strands = strands(s);
		maxval = inf; 
		if ~iscell(CFG.bam_fnames)
			gg = add_reads_from_bam(gg, CFG.bam_fnames, 'exon_track', '', maxval, CFG.read_filter, CFG.var_aware, CFG.only_primary);
		else
			for f = 1:length(CFG.bam_fnames),
				gg = add_reads_from_bam(gg, CFG.bam_fnames{f}, 'exon_track', '', maxval, CFG.read_filter, CFG.var_aware, CFG.only_primary);
			end ;
			%%%% sum of mapped_exon_tracks (odd) and spliced exon tracks (even)
			%gg.tracks = [sum(gg.tracks(1:2:end,:),1);sum(gg.tracks(2:2:end,:),1)] ;
		end ;
		if gg.strand == '-',
			gg.tracks = gg.tracks(:, end:-1:1) ;
		end ;

        %%% add introns implied by splicegraph to the list
        all_introns = gg.introns{s};
        for k = 1:size(gg.splicegraph{2}, 1),
            for l = k + 1:size(gg.splicegraph{2}, 1),
                if gg.splicegraph{2}(k, l) == 1,
                    all_introns(:, end + 1) = [gg.splicegraph{1}(2, k) + 1; gg.splicegraph{1}(1, l) - 1];
                end;
            end;
        end;
        all_introns = unique(all_introns', 'rows')';
   
        %%% use only relevant introns (inside gene boundaries)
        if ~isempty(all_introns),
            rm_idx = all_introns(2, :) <= gg.start | all_introns(1, :) >= gg.stop;
            all_introns(:, rm_idx) = [];
            clear rm_idx;
        end;

        segment_bounds = sort(unique(all_introns(:)));

		%%% check for all intron-pairs, if exon could exist between them
		new_cassette = zeros(length(all_introns)) ;
        for k = 1:size(all_introns, 2),
            for l = k + 1:size(all_introns, 2),
                if all_introns(2, k) >= all_introns(1, l),
                    continue;
                end;
                %%% only take intron pair, if outer ends are supported by current exons
                if isempty(find(gg.splicegraph{1}(2, :) == all_introns(1, k) - 1)) || isempty(find(gg.splicegraph{1}(1, :) == all_introns(2, l) + 1)),
                    continue;
                end;
                curr_exon = [all_introns(2, k) + 1, all_introns(1, l) - 1];
                %%% do not allow curr_exon to overlap existing exon
                if ~isempty(find(gg.splicegraph{1}(1, :) < curr_exon(2) & gg.splicegraph{1}(2, :) > curr_exon(1)))
                    continue;
                end;

                num_exons = num_exons + 1;

                if ~ismember(curr_exon, gg.splicegraph{1}', 'rows'),
                    idx = [curr_exon(1):curr_exon(2)] - gg.start + 1;
                    exon_cov = sum(gg.tracks(:, idx), 1);
                    pre_segment_end = find(segment_bounds < curr_exon(1) - 1, 1, 'last');
                    if ~isempty(pre_segment_end),
                        pre_segment_cov = sum(gg.tracks(:, [segment_bounds(pre_segment_end) : curr_exon(1) - 1] - gg.start + 1), 1);
                    else
                        pre_segment_cov = sum(gg.tracks(:, 1 : curr_exon(1) - gg.start), 1);
                    end;
                    min_len_pre = min(length(pre_segment_cov), length(exon_cov));

                    aft_segment_start = find(segment_bounds > curr_exon(2) + 1, 1, 'first');
                    if ~isempty(aft_segment_start),
                        aft_segment_cov = sum(gg.tracks(:, [curr_exon(2) + 1 : segment_bounds(aft_segment_start) ] - gg.start + 1), 1);
                    else
                        aft_segment_cov = sum(gg.tracks(:, curr_exon(2) + 1 - gg.start + 1 : end), 1);
                    end;
                    min_len_aft = min(length(aft_segment_cov), length(exon_cov));
                    if mean(exon_cov > 0.2*mean(exon_cov)) > CFG.cassette_exon.min_cassette_region && ...
                       median(exon_cov) > CFG.cassette_exon.min_cassette_cov && ...
                       max(median(exon_cov(end-min_len_aft+1:end)), median(aft_segment_cov(1:min_len_aft))) / min(median(exon_cov(end-min_len_aft+1:end)), median(aft_segment_cov(1:min_len_aft))) - 1 >= CFG.cassette_exon.min_cassette_rel_diff && ...
                       max(median(exon_cov(1:min_len_pre)), median(pre_segment_cov(end-min_len_pre+1:end))) / min(median(exon_cov(1:min_len_pre)), median(pre_segment_cov(end-min_len_pre+1:end))) - 1 >= CFG.cassette_exon.min_cassette_rel_diff,
                        new_cassette(k, l) = 1;
                        inserted.cassette_exon = inserted.cassette_exon + 1;
                    end;
                end;
            end;
        end;
        any_added = 0 ;
		if any(new_cassette(:))
            curr_sg = gg.splicegraph{1};
            for k = 1:size(new_cassette, 2),
                for l = k + 1:size(new_cassette, 2),
                    if new_cassette(k,l) > 0,
                        exons_pre = find(curr_sg(2, :) == (all_introns(1, k) - 1));
                        exons_aft = find(curr_sg(1, :) == (all_introns(2, l) + 1));

                        gg.splicegraph = add_cassette_exon(gg.splicegraph, [all_introns(2, k) + 1, all_introns(1, l) - 1], exons_pre, exons_aft) ;
                        new_cassette(k,l) = 0 ;
                        any_added = 1 ;
                        num_exons_added = num_exons_added + 1 ;
                    end ;
                end ;
            end ;
            [dummy,exon_order] = sort(gg.splicegraph{1}(1,:),2,'ascend');
            gg.splicegraph{1} = gg.splicegraph{1}(:,exon_order);
            gg.splicegraph{2} = gg.splicegraph{2}(exon_order,exon_order) ;
            if length(gg.splicegraph)>2
                gg.splicegraph{3} = gg.splicegraph{3}(:, exon_order) ;
            end;
            if ~any_added, break ; end ;
		end ;
		if any_added,
			[dummy,exon_order] = sort(gg.splicegraph{1}(1,:),2,'ascend');
			gg.splicegraph{1} = gg.splicegraph{1}(:,exon_order);
			gg.splicegraph{2} = gg.splicegraph{2}(exon_order,exon_order) ;
			if length(gg.splicegraph)>2
				gg.splicegraph{3} = gg.splicegraph{3}(:, exon_order) ;
			end;
		end ;
		if rm_strands
			gg = rmfield(gg, 'strands') ;
		end
        %%% clean up gene structure
		gg = rmfield(gg, 'tracks') ;
		genes(chunk_idx(c)) = gg ;
		c = c + 1;
	end;
end;
%[num_exons num_exons_added]

File: src/utils/insert_cassette_exons.py
# written by Andre Kahles, Mpi Tuebingen, Germany, 2012

def insert_cassette_exons(genes, CFG):
    # [genes, inserted] = insert_cassette_exons(genes, CFG)

    inserted['cassette_exon'] = 0

    ### form chunks for quick sorting
    chunks = sp.c_[sp.array(genes.chr_num, dtype = 'int'), sp.array(genes.strand, dtype = 'int'), sp.array(genes.start, dtype = 'int'), sp.array(genes.stop. dtyp = 'int')]
    (chunks, chunk_idx) = sort_rows(chunks, index=True)

    strands = ['+', '-']

    ### form all possible combinations of contigs and strands --> regions
    regions = init_regions(CFG['bam_fnames'])
    ### keep only chromosomes found in genes
    keepidx = sp.where(sp.in1d(sp.arra([x.chr_num for x in regions]), sp.unique(sp.array([x.chr_num for x in genes]))))[0]
    regions = regions[keepidx]

    c = 0
    num_exons_added = 0
    num_exons = 0

    for j in range(regions.shape[0]):
        chr_num = regions[j].chr_num
        s = strands.index(regions[j].strand)
        
        # fill the chunks on the corresponding chromosome
        while c <= chunks.shape[0]:
            if chunks[c, 0] > chr_num or chunks[c, 1] > strands[s]:
                break
            if chunks[c, 0] != chr_num:
                print >> sys.stderr, 'ERROR: c logic seems wrong'
                sys.exit(1)

            if CFG['verbose'] and c % 100 == 0:
                print '\r %i(%i) genes done (found %i new cassette exons in %i tested intron pairs, %2.1f%%)' % (c, chunks.shape[0], num_exons_added, num_exons, 100*num_exons_added/float(num_exons))

            gg = genes[chunk_idx[c]]
            gg.strand = strands[s]
            tracks = add_reads_from_bam(gg, CFG['bam_fnames'], ['exon_track'], CFG['read_filter'], CFG['var_aware'])

            ### add introns implied by splicegraph to the list
            all_introns = gg.introns[s]
            for k in range(gg.splicegraph.edges.shape[0]):
                for l = range(k+1, gg.splicegraph.edges.shape[0]):
                    if gg.splicegraph.edges[k, l] == 1:
                        all_introns = sp.r_[all_introns, sp.array([gg.splicegraph.vertices[1, k], gg.splicegraph.vertices[0, l]])] # introns are half open
            all_introns = unique_rows(all_introns)
       
            ### use only relevant introns (inside gene boundaries)
            if all_introns.shape[0] > 0:
                keep_idx = sp.where((all_introns[:, 1] > gg.start) & (all_introns[:, 0] < gg.stop))[0]
                all_introns = all_introns[keep_idx, :]

            segment_starts = sp.sort(sp.unique(all_introns[:, 0]))
            segment_ends = sp.sort(sp.unique(all_introns[:, 1]))

            ### check for all intron-pairs, if exon could exist between them
            new_cassette = sp.zeros((all_introns.shape[0],)) 
            for k in range(all_introns.shape[0]):
                for l in range(k + 1, all_introns.shape[0]):
                    if all_introns[k, 1] >= all_introns[l, 0]:
                        continue
                    ### only take intron pair, if outer ends are supported by current exons
                    if (not all_introns[k, 0] in gg.splicegraph.vertices[1, :]) or (not all_introns[l, 1] in gg.splicegraph.vertices[0, :]): 
                        continue
                    curr_exon = [all_introns[k, 1], all_introns[l, 0]]
                    ### do not allow curr_exon to overlap existing exon
                    if sp.sum((gg.splicegraph.vertices[0, :] < curr_exon[1]) & (gg.splicegraph.vertices[1, :] > curr_exon[0])) == 0:
                        continue

                    num_exons += 1

                    if not ismember(curr_exon, gg.splicegraph.vertices.T, rows=True):
                        idx = sp.arange(curr_exon[0], curr_exon[1]) - gg.start
                        exon_cov = sp.sum(tracks[:, idx], axis=0)

                        pre_segment_end = sp.where(segment_ends < curr_exon[0])[0].max()
                        if pre_segment_end.shape[0] > 0:
                            pre_segment_cov = sp.sum(tracks[:, sp.arange(segment_ends[pre_segment_end], curr_exon[0]) - gg.start], axis=0)
                        else:
                            pre_segment_cov = sp.sum(tracks[:, sp.arange(curr_exon[0] - gg.start)], axis=0)
                        min_len_pre = min(pre_segment_cov.shape[0], exon_cov.shape[0])

                        aft_segment_start = sp.where(segment_starts > curr_exon[1])[0].min()
                        if aft_segment_start.shape[0] > 0:
                            aft_segment_cov = sp.sum(tracks[:, sp.arange(curr_exon[1], segment_starts[aft_segment_start]) - gg.start], axis=0)
                        else:
                            aft_segment_cov = sp.sum(tracks[:, (curr_exon[1] - gg.start):], axis=0)
                        min_len_aft = min(aft_segment_cov.shape[0], exon_cov.shape[0])

                        if sp.mean(exon_cov > (0.2 * sp.mean(exon_cov))) > CFG['cassette_exon']['min_cassette_region'] and
                           sp.median(exon_cov) > CFG['cassette_exon']['min_cassette_cov'] and
                           max(sp.median(exon_cov[-min_len_aft:]), sp.median(aft_segment_cov[:min_len_aft])) / min(sp.median(exon_cov[-min_len_aft:]), sp.median(aft_segment_cov[:min_len_aft])) - 1 >= CFG['cassette_exon']['min_cassette_rel_diff'] and
                           max(sp.median(exon_cov[:min_len_pre]), sp.median(pre_segment_cov[-min_len_pre:])) / min(sp.median(exon_cov[:min_len_pre]), sp.median(pre_segment_cov[-min_len_pre:])) - 1 >= CFG['cassette_exon']['min_cassette_rel_diff']:
                            new_cassette[k, l] = 1
                            inserted['cassette_exon'] += 1 
            any_added = False
            if any(new_cassette.ravel()):
                curr_sg = gg.splicegraph.vertices
                for k in range(new_cassette.shape[1]):
                    for l = range(k + 1, new_cassette.shape[1]):
                        if new_cassette[k, l] > 0:
                            exons_pre = sp.where(curr_sg[1, :] == all_introns[k, 0])[0]
                            exons_aft = sp.where(curr_sg[0, :] == all_introns[l, 1])[0]

                            gg.splicegraph.add_cassette_exon(sp.array([all_introns[k, 1], all_introns[l, 0]]), exons_pre, exons_aft)
                            new_cassette[k, l] = 0
                            any_added = True
                            num_exons_added += 1
                exon_order = sp.argsort(gg.splicegraph.vertices[0, :])
                gg.splicegraph.reorder(exon_order)
                if not any_added:
                    break
            if any_added:
                exon_order = sp.argsort(gg.splicegraph.vertices[0, :])
                gg.splicegraph.reorder(exon_order)
            ### clean up gene structure
            genes[chunk_idx[c]] = gg
            c += 1
